### DISCUSSION TITLE
HOWTO: Run your own multiplayer server

### DISCUSSION THREAD

**Discussion by MrStahlfelge** [MrStahlfelge/lightblocks#26] [Show and tell]
For the new multiplayer feature, it is needed to connect to a server to play against the bot or others. You can connect to a public server to play, but there might be some disadvantages: You need a good and reliable connection to the server, and you don't have a private space to play against your friends only.

Running your own server is easy. It is comparable to running an own Minecraft server. If you are familiar with virtual servers and docker, you can run it on dokku from source.

## Running the server locally
The server is a Java application. You can download a precompiled jar on Linux, Windows, MacOS, Raspian from the [Releases page](https://github.com/MrStahlfelge/lightblocks/releases) here on GitHub. You need a Java Runtime Environment or JDK8 installed on the machine. For Windows, you can download one from [AdoptOpenJDK](https://adoptopenjdk.net/)

Run the server with `java -jar server.jar`. It will output some hints on more arguments you can use, write an XML file you can edit to change some settings, but most importantly, it will output if the server is up and running. When it is, you will be able to see your server in Falling Lightblocks "Local servers..." list when you are connected to the same network the server is running in. If not, you can probably connect manually by using the IP address the server output shows you.

## Make the server public
If you run the server on one of your local computers, you can only connect to it from within your own network. That's okay for local play with your siblings and family. If you want to invite a friend to play with you, you need to open the server's port for access from the internet. The port the server is using is written out when the server started. Follow guides on how to open or forward a port for public use. There are plenty of guides regarding Minecraft servers, and it is exectly the same for Lightblocks' server. When you've done so, your friends will be able to connect to the server when knowing your public IP address and using the "Enter IP" button.

## Run the server on an Internet server
You can also run the server on an Internet cloud server. I can recommend to use Docker to do so. This repository is preconfigured to run the server on [Dokku](http://dokku.viewdocs.io/dokku/).

## Add the server to the official list
If you are the proud owner or administrator of a reliable cloud server and are willing to contribute to the project by allowing every player to use it, let me know. I would be glad to add the server to Falling Lightblock's server list. At the moment, we only have one strong server in Europe. It would be great to have a server for every continent for players to use.

**Comment by totallyphilip** [comment]
I'm looking for more detailed documentation on runtime parameters. I set up the server.xml file, but my server name is still showing up in Android as the default "Lightblocks Server", so I think I'm doing something wrong.

**Reply by MrStahlfelge** [reply]
No, you did nothing wrong, but found a bug. It is fixed with the latest commit, if you are able to you can build the latest version, otherwise the problem is gone with the next release.

Thanks for reporting!

**Reply by totallyphilip** [reply]
Oh good! I was just coming back to clarify that once I connect to a game, I see the correct information. The incorrect name only displays in the lobby where all available servers are listed.

**Reply by ChaosMarc** [reply]
just in case someone else wants to create the mentioned server.xml but has no clue (like me) how it has to be structured. this is a working blueprint:

```properties
<!DOCTYPE properties SYSTEM "http://java.sun.com/dtd/properties.dtd">
<properties>
        <entry key="name">Server Name</entry>
        <entry key="owner">Server Owner</entry>
        <entry key="description">Server Description</entry>
		<!-- Not sure what this does, but the default value is false -->
        <entry key="privateRooms">false</entry>
		<!-- Classic = 0, Modern = 1, Both = 2 -->
        <entry key="gamemodes">0</entry>
</properties>
```

**Reply by MrStahlfelge** [reply]
Private Rooms is planned for people to create rooms that are not public but only accessible with user/pass
